[CALCITE-3272] Support TUMBLE as Table Valued Function including an enumerable implementation, stream.iq and DESCRIPTOR

amaliujia opened the pull request:
High level speaking, this PR adds the following support:

```
SELECT * FROM TABLE(Tumble( TABLE ORDERS , DESCRIPTOR(ROWTIME) , INTERVAL '1' MINUTES))
```


This PR adds TUMBLE as table value function and also adds stream.iq along with Enumerable implementation. This is a big PR that actually is also related to the following JIRAs:

https://jira.apache.org/jira/browse/CALCITE-3340
https://jira.apache.org/jira/browse/CALCITE-3501
https://jira.apache.org/jira/browse/CALCITE-3499
https://jira.apache.org/jira/browse/CALCITE-3418
https://jira.apache.org/jira/browse/CALCITE-3339


Note that **DESCRIPTOR** support is also included in this PR.

amaliujia commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java:
```diff
@@ -940,6 +956,49 @@ public static Expression translateLiteral(
+      translatedOperandsForTumble.add(translatedOperands.get(watermarkedColumnIndex));
+      translatedOperandsForTumble.add(intervalExpression);
+      // compute the window_start
+      MethodNameImplementor windowStartMethod = new MethodNameImplementor("tumbleWindowStart");
```
I found it was hard to add these to `RexImpTable` because one operator has two implementor: window_start and window_end.

For TUMBLE and HOP, a workaround is only provide implementor for window_start and later compute window_end by window_start + window_size. However, it won't work for `SESSION`.

amaliujia commented on core/src/main/java/org/apache/calcite/sql/SqlKind.java:
```diff
@@ -988,17 +988,21 @@
+  DESCRIPTOR,
 
-  /** The {@code TUMBLE} group function. */
+  /** The {@code TUMBLE_TABLE_FUNCTION}. */
```
I thought Sqlkind.TUMBLE will be dedicated to the built-in table-valued function. Given another thought now, since the old TUMBLE is not deprecated yet, so I roll back this change to make SqlKind.TUMBLE still be used to refer the old TUMBLE.

We can change all the comments together when be ready to deprecate old TUMBLE.

julianhyde replied:
My point was that TUMBLE_TABLE_FUNCTION is not a SQL keyword, so it shouldn't be "code".

Ideally TUMBLE becomes an ordinary table function. If so, we can stop using SqlKind.TUMBLE as a special case. If not, that's also fine.

amaliujia commented on core/src/main/java/org/apache/calcite/sql/SqlKind.java:
```diff
@@ -988,17 +988,21 @@
   CUME_DIST,
 
-  // Group functions
+  /** The {@code DESCRIPTOR}. */
```
ah. Updated the content of @code by the syntax of DESCRIPTOR.

amaliujia commented on core/src/main/java/org/apache/calcite/sql/type/TableFunctionWindowingOrdinalReturnTypeInference.java:
```diff
@@ -0,0 +1,63 @@
+ *  1. wstart. TIMESTAMP type to indicate a window's start.
+ *  2. wend. TIMESTAMP type to indicate a window's end.
+ */
+public class TableFunctionWindowingOrdinalReturnTypeInference implements SqlReturnTypeInference {
```
Changed to anonymous inner class

amaliujia commented on core/src/test/java/org/apache/calcite/tools/PlannerTest.java:
```diff
@@ -254,7 +255,7 @@ private Planner getPlanner(List<RelTraitDef> traitDefs,
         .parserConfig(parserConfig)
         .defaultSchema(
-            CalciteAssert.addSchema(rootSchema, CalciteAssert.SchemaSpec.HR))
+            CalciteAssert.addSchema(rootSchema, SchemaSpec.HR))
```
Changed it back.

amaliujia commented on core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:
```diff
@@ -2455,7 +2455,13 @@ protected void convertCollectionTable(
-    final List<RelNode> inputs = bb.retrieveCursors();
+    List<RelNode> inputs = new ArrayList<>();
+
+    if (rexCall.getKind() == SqlKind.TUMBLE) {
```
I found I need this line to make `TABLE` parameter supported. Otherwise the inputs to TUMBLE table function will be empty even though there is a `TABLE` parameter and parser successfully parses it to a `SqlSelect`.

julianhyde replied:
It's possible that the fix is in `bb.retrieveCursors()`. Also possible, and in fact better, would be to let Calcite treat `TUMBLE` as an ordinary table function (in other words, don't check for `SqlKind.TUMBLE`). Then everything might just work out of the box.

amaliujia replied:
I removed this check here and give it a try: confirmed again Inputs will be empty (e.g. TABLE parameter is not converted). Am I doing right to let treat `TUMBLE` as an ordinary table function? Is it actually other place to not check TUMBLE to make it as a ordinary table function?

amaliujia replied:
@julianhyde do you have more comments regarding to this line of change?

Also generally I will try to push this change into 1.22.0 release. Could you let me know if there are other big refactors that I can do so I can do them soon?

Or if you think it is less mature to let this change into 1.22.0, please let me know and it's ok. I will work toward to make it released in 1.23.0

danny0405 replied:
In order to hack into this method, we may match the `TUMBLE` table value function just like lateral table [1]

[1] https://github.com/apache/calcite/blob/b93ec5a9edb7459696385e6adad67b92b6d406d7/core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java#L2199

amaliujia replied:
I changed the `SqlKind` of TUMBLE table-valued funciton to `COLLECTION_TABLE`. It will have the same problem that the input of `TableFunctionScan` is empty, so basically it does not fix the `bb.retrieveCursors();`

amaliujia commented on core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:
```diff
@@ -2739,7 +2740,25 @@ private void registerQuery(
       registerOperandSubQueries(parentScope, call, 0);
       scopes.put(node, parentScope);
       break;
+    case TUMBLE:
```
Existing table functions' namespace building is top-down. In order to support DESCRIPTOR for TUMBLE, we need a bottom-up approach in which
1. build scope of TABLE parameter first (which is a SqlSelect). 
2. build namespace and use the scope of step 1 for TUMBLE.

By doing, in the following validation steps, TUMBLE will have a right scope to allow it's operand (the descriptor) to solve the name. 

If not doing this, the descriptor will only see an EmptyScope.

amaliujia commented:
@julianhyde thanks for reviewing. Have tried to address your comments.

danny0405 commented on core/src/main/java/org/apache/calcite/sql/SqlTableValuedFunctionWindowingOperator.java:
```diff
@@ -0,0 +1,48 @@
+      String name, SqlKind kind,
+      SqlReturnTypeInference returnTypeInference,
+      SqlOperandTypeInference operandTypeInference,
+      SqlOperandTypeChecker operandTypeChecker,
```
Fix the indentation.

danny0405 commented:
Thanks @amaliujia  for firing this PR, i'm planning to give a review and can you resolve the conflicts ?

amaliujia commented:
@danny0405 Thank you!

conflict solved and indentation fixed.

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java:
```diff
@@ -937,6 +950,22 @@ public static Expression translateLiteral(
+  private List<Expression> translateTableFunctionCall(RexCall rexCall) {
+    if (rexCall.op.getKind() == SqlKind.TUMBLE) {
+      assert rexCall.getOperands().get(2) instanceof RexLiteral;
+      Expression intervalExpression = translate(rexCall.getOperands().get(2));
```
We may firstly assert that the operands number is correct to avoid `INDEX OUT OF BOUNDES`. We may also give some messages to these assertions.

amaliujia replied:
Will do.

amaliujia replied:
addressed

danny0405 commented on core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:
```diff
@@ -2285,6 +2285,7 @@ private SqlNode registerFrom(
     case VALUES:
     case WITH:
     case OTHER_FUNCTION:
+    case TUMBLE:
```
This is not necessary, handle the function like lateral table [1]

[1] https://github.com/apache/calcite/blob/b93ec5a9edb7459696385e6adad67b92b6d406d7/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java#L2256

amaliujia replied:
I believe in Calcite `COLLECTION_TABLE` is the `TABLE(...)` part in `SELECT * FROM TABLE(TUMBLE(...))`. 

I will give it try to change TUMBLE's sql kind to `COLLECTION_TABLE` tomorrow and see how will it work.

amaliujia replied:
Ok. So I think the suggestion to use `COLLECITON_TABLE` is to address the special handling in validator. I gave it a try to make `TUMBLE`'s SqlKind as `COLLECITON_TABLE` and found

1. most of the special handling for TUMBLE is not needed
2. except for 
```
    if (rexCall.getKind() == SqlKind.TUMBLE) {
      inputs.add(convertSelect((SqlSelect) call.getOperandList().get(0), false));
    } else {
      inputs = bb.retrieveCursors();
    }
```
as inputs is still empty by using `COLLECITON_TABLE`.

3. In enumerable implementation, string match is used to check if it is tumbling case as we no longer can use SqlKind to check it, which makes code a little longer.


Because the inputs is still empty problem, I incline to still use SqlKind.TUMBLE.

amaliujia commented on core/src/main/java/org/apache/calcite/sql/SqlTableValuedFunctionWindowingOperator.java:
```diff
@@ -0,0 +1,48 @@
+ * Base class for table-valued function windowing operator (TUMBLE, HOP and SESSION).
+ */
+public class SqlTableValuedFunctionWindowingOperator extends SqlFunction {
+  public SqlTableValuedFunctionWindowingOperator(String name,
```
Gotcha. Currently this PR is only supports `TUMBLE`, so we can make `TUMBLE to extends SqlFunction` and override `argumentMustBeScalar`. 

The plan is I will also add support for HOP(sliding windowing) and SESSION (sessionization), and those will also need to override `argumentMustBeScalar`, thus a base class will make sense for that time (and I just implement it in this PR).

amaliujia commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java:
```diff
@@ -937,6 +950,22 @@ public static Expression translateLiteral(
+      List<Expression> translatedOperandsForTumble = new ArrayList<>();
+      translatedOperandsForTumble.add(
+          Expressions.constant(((RexInputRef) descriptor.getOperands().get(0)).getIndex()));
+      translatedOperandsForTumble.add(intervalExpression);
```
TUMBLE's signature is `TUMBLE(table, DESCRIPTOR(column_name), interval [, time ])`, so it's supposed to only have one column name in `DESCRIPTOR` to indicate watermarked event timestamp column.

It's good question that where TUMBLE's signature is validated such that we know `DESCRIPTOR` only has one column name. The answer is seems to me that the best place is in parser, in which it will only create TUMBLE if `DESCRIPTOR` has one column. 

I am planning to leave it as a future work.

amaliujia commented on core/src/main/java/org/apache/calcite/sql/SqlDescriptorOperator.java:
```diff
@@ -0,0 +1,73 @@
+            RESOURCE.aliasMustBeSimpleIdentifier());
+      }
+
+      SqlIdentifier identifier = (SqlIdentifier) node;
```
I don't have an answer on it. 

By tracking down the code execution, this is the place I found I am able to verify what's inside `DESCRIPTOR`.

danny0405 replied:
Did you try overriding `checkOperandTypes` ?

amaliujia replied:
This is tricky for `DESCRIPTOR`. 

After checking existing type checkers, I think Calcite' `OperandTypeChecker` and `checkOperandTypes` are designed to verify the operand types. For example. It checks if MOD's operands are `INTEGER`.


Back to `DESCRIPTOR`, there is no really `type` associated with column names. For example, `descriptor(key_column)`, we don't know what's the type of key_column, thus type checker checks nothing.

There is a need verify if columns can be found, which can be done in either `deriveType` or `checkOperandTypes`. 

In the full version of  `DESCRIPTOR`, SQL standards allows user provide a type, e.g. `descriptor(key_column INTEGER)`, which can be used for type checking, say the column name can be resolved and the corresponding column found from the table is also `INTEGER`. The full version should override `checkOperandTypes` for sure.

Given the current status of `DESCRIPTOR`, I incline to keep the name resolution in `driveType` as it's already implemented. When we have the full version of `DESCRIPTOR`, we should make everything done in  type checker.

amaliujia replied:
OK. Had a second thought and I finally move column name resolution to the operand checker. 

Like said `DESCRIPTOR` is not mature enough and it will need evolve for sure.

amaliujia commented on core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:
```diff
@@ -510,6 +510,52 @@ public boolean checkOperandTypes(
+
+        public SqlOperandCountRange getOperandCountRange() {
+          return SqlOperandCountRanges.of(3);
+        }
```
It's used by `TUMBLE`, which has three arguments. Is there a suggestion how to set such numbers?

danny0405 replied:
But you name it `ROW`, why a row operand type must have 3 operands ?

amaliujia replied:
Now I finally understand how a right `OperandTypeChecker` should be written. I have update this `ROW` to `TABLE_VALUED_FUNCTION_TUMBLE_CHECKER` and add more comments. Now this checker checks TUMBLE only with 3 operand, whose types are `ROW`, `COLUMN_LIST`, `DATETIME_INTERVAL` separately.  Please take a look.

danny0405 replied:
Move this operand type check logic into `SqlTableValuedFunctionWindowingOperator` directly.

amaliujia replied:
done

amaliujia commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableTableFunctionScan.java:
```diff
@@ -61,28 +78,32 @@ public EnumerableTableFunctionScan(RelOptCluster cluster,
+      return tableValuedFunctionWindowingImplement(implementor, pref);
     } else {
-      format = JavaRowFormat.CUSTOM;
+      BlockBuilder bb = new BlockBuilder();
```
done

amaliujia commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -2706,6 +2706,14 @@ public static int subtractMonths(long t0, long t1) {
   }
 
+  public static long tumbleWindowStart(long tsMillis, long sizeMillis) {
+    return tsMillis / sizeMillis * sizeMillis;
```
done

amaliujia commented:
Checks are failing because of [CALCITE-3604](https://issues.apache.org/jira/browse/CALCITE-3604)

amaliujia commented:
@danny0405 

Ok, until this moment I have tried addressed all your comments. Especially I have made changes to make that `ROW` checker not confusing anymore. All checks are also green now.

Could you take a another look?

danny0405 commented on core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:
```diff
@@ -90,6 +93,33 @@ public static ExplicitReturnTypeInference explicit(SqlTypeName typeName,
+            new RelDataTypeFieldImpl("window_start", newFields.size(), timestampType);
+        newFields.add(windowStartField);
+        RelDataTypeField windowEndField =
+            new RelDataTypeFieldImpl("window_end", newFields.size(), timestampType);
```
Move this return type check logic into `SqlTableValuedFunctionWindowingOperator` directly.

amaliujia replied:
done

amaliujia commented:
A friendly ping~ comments were addressed.

@danny0405 I really appreciate your review on this PR. It's very important for BeamSQL! Thanks a lot.

danny0405 commented:
I tried to modify based on your PR to treat `Tumble` as a normal table value function: use `SqlKind.OTHER_FUNCTION` to make the code works as the same, still in the progress ~

During the working, i have a feeling that the `Descriptor` should not have a different scope with the table value function, in the table value function's first operand scope, we should have see the referenced table columns.

We may also need to add more tests:

- For SqlParserTest: to test all the syntax with positive/negative cases
- For SqlValidatorTest: test you operands/return type inference defined in the 2 new operators
- For SqlToRelConverterTest: test your plan is correct after the conversion

amaliujia commented:
Gotcha. So you think the better way is to use `SqlKind.OTHER_FUNCTION`? It's fine to me as I don't know what's the better solution for that.

`Descriptor` is supposed to have the same scope as the TVF for sure. I think that's what I am doing this PR: build the scope from the first operand of TVF, then use that scope for `Descriptor`.

I will try to add more tests first then. That should be easier to do.

danny0405 commented:
Remove the unnecessary changes in `OperandTypes` and `ReturnTypes`.

amaliujia commented:
just Remove the unnecessary changes in OperandTypes and ReturnTypes.

- For SqlParserTest. Added a few more tests related to `DESCRIPTOR`.  There is no obvious parser negative syntax so didn't add any. Table functions are already tested so didn't add any either.
- For SqlValidatorTest. Tested incorrect signatures of `TUMBLE`.
- SqlToRelConverterTest. Added a test to test generated plan.

danny0405 commented:
Here is my re-work based on your code, https://github.com/danny0405/calcite/tree/amaliujia-rw-add_TVF_TUMBLE :
- Fix the logic in SqlValidator and SqlToRelConverter
- Fix the code generation

There are still 2 TODOs left there, hope you can solve that :)

amaliujia commented:
@danny0405 thanks for your awesome work! I merged your change into this PR. I tried to squash this PR to two commits: one is my change and then one is yours so we can keep your credit in the git log. Unfortunately I messed that up and all the changes are merged into one single commit now.

Please let me know if you want to have two commits to separate changes. I can try to rebuild and see if I can recover it.

I was able to address one of the TODO. I didn't figure out this TODO:
```
    if (isTumble) {
      // TODO: should move this code to RexImpTable and keep the code
      // in this class unchanged.
      return tableValuedFunctionWindowingImplement(implementor, pref);
```

I was guessing you want to use [1] to encapsulate some of the code into `RexImpTable`. As far as I can tell, `CallImplementor`'s interface is a little simple to encapsulate all the enumerable code of tumbling. I checked what is the existing practice. For example, I checked `EnumerableMergeJoin`, and `EnumerableMergeJoin` has the same practice as what is doing in `EnumerableTableFunctionScan`.

[1]: https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/adapter/enumerable/CallImplementor.java#L38

danny0405 commented:
No need to keep my credit, the `if (isTumble)` decision is too hacky for me, so we must solve that.

amaliujia commented:
@danny0405 
I understand your concerns. Current implementation is less extensible as there will be more built-in table-valued function implementations and we cannot have many if-else with their own enumerable implementations.  An implementor interface would be desired for each implementation. 

However, I believe current `CallImplementor` is not enough for table-valued functions. Table-valued functions cannot be done by the following code because at least it requires `inputPhysType`, and very likely require `outputPhysType`and others params.
```
// RexCalls that are suitable for `CallImplementor`
for (operand : call.operands())  {
   translate(operand);
}

translate(call, translatedOperands);
```  

I think we need a new interface for this purpose, I have filed [CALCITE-3629](https://issues.apache.org/jira/browse/CALCITE-3629). 

In this PR, I made some changes to move some Tumbling code to `EnumUtils`, per the existing practice of enumerables and not every line of code can be moved out. Meanwhile, I would suggest to solve [CALCITE-3629](https://issues.apache.org/jira/browse/CALCITE-3629) when I add `HOP` implementation (a work after this PR). The rationale is we need more implementations to build an abstraction and then which leads to a good interface design.

amaliujia commented:
Ok. I added CALCITE-3629 as a separate commit in this PR, in which a new implementor interface for table-valued functions is added. Some of the implementations of tumbling is moved to `RexTableImpl`. The new interface is marked as experimental because it could change in the future as new builtin table-valued functions are added.

Please let me know if CALCITE-3629 should be in a separate PR. I am happy to split it.

danny0405 commented on core/src/main/java/org/apache/calcite/sql/SqlWindowTableFunction.java:
```diff
@@ -0,0 +1,129 @@
+
+  @Override public boolean checkOperandTypes(SqlCallBinding callBinding,
+                                             boolean throwOnFailure) {
+    // There should only be three operands, and number of operands are checked before
```
Fix the indention.

danny0405 commented on core/src/main/java/org/apache/calcite/sql/SqlWindowTableFunction.java:
```diff
@@ -0,0 +1,129 @@
+      boolean matches = false;
+      for (String field : type.getFieldNames()) {
+        if (validator.getCatalogReader().nameMatcher().matches(field, colName)) {
+          matches = true;
```
Actually the name validation expects to happen in `SqlValidator`, but now the logic is simple, putting it here can make the `SqlValidator` unchanged.

danny0405 commented on core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:
```diff
@@ -4783,8 +4783,22 @@ GROUP BY TUMBLE(rowtime, INTERVAL '1' MINUTE)]]>
+            <![CDATA[
+LogicalProject(ORDERID=[$0], ROWTIME=[$1], window_start=[$2], window_end=[$3])
+  LogicalTableFunctionScan(invocation=[TUMBLE($1, DESCRIPTOR($1), 60000:INTERVAL MINUTE)], rowType=[RecordType(INTEGER ORDERID, TIMESTAMP(0) ROWTIME, TIMESTAMP(0) window_start, TIMESTAMP(0) window_end)])
+    LogicalProject(ORDERID=[$0], ROWTIME=[$1])
```
After conversion, the `TUMBLE` call first operand is always last field of it's input, which need a note because in current relational algebra, there is no way to give the first operand a proper type.

amaliujia replied:
Added notes to related tests.

amaliujia commented:
@danny0405 

Thanks Danny! I added two more tests in both `stream.iq` and `SqlToRelConverterTest` with using `subquery` as the first operand of `TUMBLE` to demonstrate that it is still functional in that case.

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:
```diff
@@ -293,6 +295,8 @@
   private final Map<SqlMatchFunction, Supplier<? extends MatchImplementor>> matchMap =
       new HashMap<>();
+  private final Map<SqlOperator, TableValuedFunctionCallImplementor> tvfImplementorMap =
+      new HashMap<>();
```
The value type should also be a `Supplier`.

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java:
```diff
@@ -637,4 +638,57 @@ static Expression generatePredicate(
   }
+
+  /** Generates a window selector which appends attribute of the window bases on
+   * the parameters. */
```
`based on`

amaliujia replied:
ah thanks!

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:
```diff
@@ -928,6 +933,10 @@ public MatchImplementor get(final SqlMatchFunction function) {
   }
 
+  public TableValuedFunctionCallImplementor getTvfImplementor(final SqlOperator operator) {
+    return tvfImplementorMap.get(operator);
```
Rename this method to `get` and pass in argument as type `SqlWindowTableFunction`.

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/TableValuedFunctionCallImplementor.java:
```diff
@@ -0,0 +1,43 @@
+   * @param inputPhysType Physical type of the table parameter
+   * @param outputPhysType Physical type of the call
+   * @return
+   */
```
Give every annotation a comment. Start with uppercase character and end with dot.

amaliujia commented:
Comments addressed.

amaliujia commented:
The failed tests seems not relevant to this PR:

```
org.apache.calcite.adapter.elasticsearch.ScrollingTest STANDARD_OUT
    [WARN ] 00:46:39.492 org.elasticsearch.cluster.service.ClusterApplierService.lambda$callClusterStateListeners$6() - failed to notify ClusterStateListener
```

amaliujia commented:
All green now.

amaliujia commented:
A friendly ping~

amaliujia commented:
@danny0405 I tried to send an email to @julianhyde to remind the status of this PR, but Julian might just be busy recently. (I also tried to ask for help from other people in community but people generally is busy recently)

So if there is no response on this PR, could we just merge it next week? I can send an email to community as a reminder a few days before we do the merging.

danny0405 commented:
Thanks for the kindly reminder @amaliujia , i still want Julian to take a look, let's wait a few days.

@julianhyde Would you please give a help here, to see if the latest code is as your expected, i always respected your suggestions ~

DonnyZone commented:
+1
LGTM

34venu commented:
Is there any calcite connection to run this tumble function
Because we are running this function in windows 7 operating system through sqlline.
As sqlline.bat
sqlline>!connect jdbc:calcite: admin admin

amaliujia commented:
@34venu 

my understanding is it should just work by calcite connection(the stream.iq actually runs in similar case). 

Have you tried to use `sqline` to try `TUMBLE` query based on head of master?

34venu commented:
I have run the tumble query(table function) with calcite connection but i
am getting the following exception.

On Tue, Jan 21, 2020 at 11:42 PM Rui Wang <notifications@github.com> wrote:

> @34venu <https://github.com/34venu>
>
> my understanding is it should just work by calcite connection(the
> stream.iq actually runs in similar case).
>
> Have you tried to use sqline to try TUMBLE query based on head of master?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/calcite/pull/1587?email_source=notifications&email_token=AJVHYGL45Y74YYL476ELSPTQ643H5A5CNFSM4JNTKOZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJQWP7Y#issuecomment-576808959>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJVHYGL7SL7W4AXZCIQHROLQ643H5ANCNFSM4JNTKOZA>
> .
>

34venu commented:
can we run tumble function through mysql or mongodb?

34venu commented:
Can we write a similar table function like tumble table function?

amaliujia commented:
@34venu 

Could you send an email to dev@calcite.apache.org about what help you need? This PR won't be a good place to help your need.

34venu commented:
@amaliujia 
can we create a similar table function like Tumble Table Function in apache calcite.Could please help me out!!
